src/node_contextify.cc

```diff
@@ -1566,6 +1568,88 @@ void ContextifyContext::ContainsModuleSyntax(
   args.GetReturnValue().Set(should_retry_as_esm);
 }
 
+void ContextifyContext::ShouldRetryAsESM(
+    const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  Local<Context> context = env->context();
+
+  if (args.Length() != 1) {
+    return THROW_ERR_MISSING_ARGS(
+        env, "shouldRetryAsESM needs 1 argument");
+  }
+  // Argument 1: source code
+  Local<String> code;
+  CHECK(args[0]->IsString());
+  code = args[0].As<String>();
+
+  Local<String> script_id = String::NewFromUtf8(isolate, "throwaway").ToLocalChecked();
+  Local<Symbol> id_symbol = Symbol::New(isolate, script_id);
+
+  Local<PrimitiveArray> host_defined_options =
+      GetHostDefinedOptions(isolate, id_symbol);
+  ScriptCompiler::Source source = GetCommonJSSourceInstance(
+      isolate, code, script_id, 0, 0, host_defined_options, nullptr);
+  ScriptCompiler::CompileOptions options = GetCompileOptions(source);
+
+  TryCatchScope try_catch(env);
+  ShouldNotAbortOnUncaughtScope no_abort_scope(env);
+
+  // Try parsing where instead of the CommonJS wrapper we use an async function
+  // wrapper. If the parse succeeds, then any CommonJS parse error for this
+  // module was caused by either a top-level declaration of one of the CommonJS
+  // module variables, or a top-level `await`.
+  code =
+      String::Concat(isolate,
+                      String::NewFromUtf8(isolate, "(async function() {")
```

Comment:
Why not just parse it as a module?

Reply:
Well, because this is how #52024 did it. Maybe it would be better to change this, but that feels unrelated to the point of this PR? I’d have to consider all the various edge cases and see if parsing this as ESM here would still catch all of them, like `const require` (might be okay, since it’s a runtime error in ESM and not a syntax error) or sloppy mode syntax (this might be an issue since it would throw a syntax error in ESM but not in CommonJS, and we _want_ this to tell us to move on to try to parse as ESM and therefore throw the ESM syntax error to the user if the “opt-in” ESM syntax is used).